### PR TITLE
i229: test auth reducers

### LINKED ISSUE #229: write reducers unit tests for auth user flows

### PR DESCRIPTION
**Pull request by tamari-gray**
fixes #229

### CODE REVIEW

File: lib/reducers/auth/store_auth_step.dart [hunk 1]
```diff
@@ -2,6 +2,7 @@ import 'package:crowdleague/actions/auth/store_auth_step.dart';
 import 'package:crowdleague/models/app/app_state.dart';
 import 'package:redux/redux.dart';
 
+/// Stores [AuthStep] that toggles UI in [AuthPage] and [EmailAuthOptionsPage]
```
Comment (tamari-gray) [inline lib/reducers/auth/store_auth_step.dart:5]: added this comment to help me understand what effects this reducer has on the app, or how its job is important in the big picture. 
happy to remove if unneccessary :)

File: lib/reducers/auth/store_user.dart [hunk 2]
```diff
@@ -5,14 +5,13 @@ import 'package:crowdleague/models/navigation/page_data/initial_page_data.dart';
 import 'package:crowdleague/models/navigation/page_data/page_data.dart';
 import 'package:redux/redux.dart';
 
-/// Set the user object to null if action has null user or update the state
```
Comment (tamari-gray) [inline lib/reducers/auth/store_user.dart:8]: user object should never be null (if it is, then clearUserData action would be dispatched) so removed nullcheck + updated comment

File: test/unit/reducers/clear_user_data_test.dart [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import 'package:crowdleague/actions/auth/clear_user_data.dart';
+import 'package:crowdleague/models/app/app_state.dart';
+import 'package:crowdleague/reducers/app_reducer.dart';
+import 'package:redux/redux.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('ClearUserDataReducer', () {
+    test('Clears any user data from state', () {
```
Comment (tamari-gray) [inline test/unit/reducers/clear_user_data_test.dart:9]: hopefully escaped the capital police 🚓

**Review by tamari-gray — COMMENTED** [review]
think shes gtg :)

File: lib/reducers/auth/store_auth_step.dart (hunk 1 above)
Comment (nickmeinhold) [reply lib/reducers/auth/store_auth_step.dart:5]: Great addition, thanks!

File: lib/reducers/auth/store_user.dart (hunk 2 above)
Comment (nickmeinhold) [reply lib/reducers/auth/store_user.dart:8]: Beautiful!

File: test/unit/reducers/clear_user_data_test.dart (hunk 3 above)
Comment (nickmeinhold) [reply test/unit/reducers/clear_user_data_test.dart:9]: Haha the police have stranger rules than you might think!  The fines are small though 😉 

You have to read b/w the lines a bit, sorry for the lack of the clarity: [Style Guide (WIP) > do-use-a-test-description-that-completes-a-sentence-starting-with](https://github.com/crowdleague/crowdleague/wiki/Style-Guide-(WIP)#do-use-a-test-description-that-completes-a-sentence-starting-with-)

So the `test('...` part should start with lower case... I feel so pedantic! 😞

File: test/unit/reducers/clear_user_data_test.dart [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      // Create a basic store with the app reducers
+      final store = Store<AppState>(
+        appReducer,
```
Comment (nickmeinhold) [inline test/unit/reducers/clear_user_data_test.dart:12]: I imagine I've been doing this myself (using the appReducer) but I think we should use just the reducer under test, if possible.  I'll see if I can make the change and add it to the PR so we can discuss tomorrow...

Comment (nickmeinhold) [reply test/unit/reducers/clear_user_data_test.dart:12]: I tried testing the reducer directly without the store and it seems to work nicely.  I put up #251 as an example. After discussing with @tamari-gray he's going to change the other test to use this method.

**Review by nickmeinhold — CHANGES_REQUESTED** [review]
Thanks mate, awesome work as usual!  

As per comments, handing back for changing the reducer tests to not use the store.

File: lib/reducers/auth/store_user.dart [hunk 5]
```diff
@@ -5,14 +5,13 @@ import 'package:crowdleague/models/navigation/page_data/initial_page_data.dart';
...
-/// with the new user.
+/// Updates state with new user
 ///
 /// We reset the pagesData list in order to pop anything on top of the main page
 /// (eg. when user has just signed in)
 class StoreUserReducer extends TypedReducer<AppState, StoreUser> {
   StoreUserReducer()
       : super((state, action) => state.rebuild((b) => b
-          ..user = action.user?.toBuilder()
+          ..user = action.user.toBuilder()
```
Comment (nickmeinhold) [inline lib/reducers/auth/store_user.dart:15]: Updated comment + removal of null check = so much clarity! 👓

File: test/unit/reducers/clear_user_data_test.dart (hunk 4 above)
Comment (tamari-gray) [reply test/unit/reducers/clear_user_data_test.dart]: discussed :)

File: lib/reducers/auth/store_user.dart (hunk 5 above)
Comment (tamari-gray) [reply lib/reducers/auth/store_user.dart:15]: 😄

File: test/unit/reducers/clear_user_data_test.dart (hunk 3 above)
Comment (tamari-gray) [reply test/unit/reducers/clear_user_data_test.dart]: no worries!

**Review by tamari-gray — COMMENTED** [review]
ready for next review 🤓

**Review by nickmeinhold — CHANGES_REQUESTED** [review]
Thanks mate!  Fair bit of nit-picky type stuff there, if you're happy to have a look and change anything you agree with / discuss anything you don't, then I'll have another pass caus I didn't really get to review with my higher level logic type brain being active.  Thanks!

File: test/unit/reducers/clear_user_data_test.dart [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      // Setup an initial app state .
+      final initialState = AppState.init();
+
+      // Create the reducer.
+      final rut = ClearUserDataReducer();
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, ClearUserData());
+
+      // Check that the new state has null for the user.
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/clear_user_data_test.dart:41]: who did this 👀

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/clear_user_data_test.dart:41]: can do :)

File: test/unit/reducers/clear_user_data_test.dart [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      expect(newState.user, null);
+    });
+
+    test('resets navigation stack', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/clear_user_data_test.dart:31]: yeep alg

File: test/unit/reducers/clear_user_data_test.dart [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      final initialState = AppState.init();
+
+      // Create the reducer.
+      final rut = ClearUserDataReducer();
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, ClearUserData());
+
+      // Check that the new state has null for the user.
+      expect(newState.pagesData, initialState.pagesData);
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/clear_user_data_test.dart:42]: on it 🖖

File: test/unit/reducers/store_auth_step_test.dart [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import 'package:crowdleague/enums/auth_step.dart';
+import 'package:crowdleague/models/app/app_state.dart';
+import 'package:crowdleague/reducers/auth/store_auth_step.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('StoreAuthStepReducer', () {
+    test('updates step for emailAuthOptionsPage', () {
+      // Setup an initial app state
+      final initialState = AppState.init();
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:11]: shweeet

File: test/unit/reducers/store_auth_step_test.dart [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+      // Create data for action
+      final testStep = AuthStep.waitingForInput;
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, StoreAuthStep(step: testStep));
+
+      // Check that the new state has the new user.
+      expect(newState.emailAuthOptionsPage.step, testStep);
+    });
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:24]: yeep 👍

File: test/unit/reducers/store_auth_step_test.dart [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import 'package:crowdleague/actions/auth/store_auth_step.dart';
+import 'package:crowdleague/enums/auth_step.dart';
+import 'package:crowdleague/models/app/app_state.dart';
+import 'package:crowdleague/reducers/auth/store_auth_step.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('StoreAuthStepReducer', () {
+    test('updates step for emailAuthOptionsPage', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:9]: salute

File: test/unit/reducers/store_auth_step_test.dart [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      // Create the reducer.
+      final rut = StoreAuthStepReducer();
+
+      // Create data for action
+      final testStep = AuthStep.waitingForInput;
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, StoreAuthStep(step: testStep));
+
+      // Check that the new state has the new user.
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:22]: shamefull copy and paste 😅

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:22]: updated :)

File: test/unit/reducers/store_auth_step_test.dart [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+      expect(newState.emailAuthOptionsPage.step, testStep);
+    });
+    test('Updates step for authPage', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:25]: im very tempted to use that john cena pic again haha

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:25]: changes added :)

File: test/unit/reducers/store_auth_step_test.dart [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+      // Setup an initial app state
+      final initialState = AppState.init();
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:27]: yeep 🏎️

File: test/unit/reducers/store_auth_step_test.dart [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+      // Create the reducer.
+      final rut = StoreAuthStepReducer();
+
+      // Create data for action
+      final testStep = AuthStep.waitingForInput;
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, StoreAuthStep(step: testStep));
+
+      // Check that the new state has the new user.
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_auth_step_test.dart:38]: updated 😅

File: test/unit/reducers/store_user_test.dart [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+      // Create the reducer.
+      final rut = StoreUserReducer();
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, StoreUser(user: mockUser));
+
+      // Check that the new state has the new user.
+      expect(newState.user, mockUser);
+    });
+    test('resets navigation stack', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:23]: done-ito 🌯

File: test/unit/reducers/store_user_test.dart [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+      // Setup an initial app state
+      final initialState = AppState.init();
+
+      // Create the reducer.
+      final rut = StoreUserReducer();
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, StoreUser(user: mockUser));
+
+      // Check that the new state has the new user.
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:20]: it was a fair assumption  😓😅😆

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:20]: oh @nickmeinhold this cracked me up

File: test/unit/reducers/store_user_test.dart [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+import 'package:crowdleague/reducers/auth/store_user.dart';
+import 'package:test/test.dart';
+
+import '../../mocks/models/user_mocks.dart';
+
+void main() {
+  group('StoreUserReducer', () {
+    test('adds new user to store', () {
+      // Setup an initial app state
+      final initialState = AppState.init();
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:12]: yeep.  i really like using expect() to document what initial values we are (cant think of another word) expecting. makes the test alot clearer!

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:12]: expect is added too :)

File: test/unit/reducers/store_user_test.dart [hunk 19]
```diff
@@ -0,0 +1,37 @@
+import 'package:crowdleague/actions/auth/store_user.dart';
+import 'package:crowdleague/models/app/app_state.dart';
+import 'package:crowdleague/reducers/auth/store_user.dart';
+import 'package:test/test.dart';
+
+import '../../mocks/models/user_mocks.dart';
+
+void main() {
+  group('StoreUserReducer', () {
+    test('adds new user to store', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:10]: sweet 🍬

File: test/unit/reducers/store_user_test.dart [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, StoreUser(user: mockUser));
+
+      // Check that the new state has the new user.
+      expect(newState.user, mockUser);
+    });
+    test('resets navigation stack', () {
+      // Setup an initial app state .
+      final initialState = AppState.init();
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:25]: yeep

File: test/unit/reducers/store_user_test.dart [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+      // Create the reducer.
+      final rut = StoreUserReducer();
+
+      // Invoke the reducer to get a new state.
+      final newState = rut.reducer(initialState, StoreUser(user: mockUser));
+
+      // Check that the new state has null for the user.
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/store_user_test.dart:33]: yeep

File: test/unit/reducers/update_email_auth_options_page_test.dart [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+import 'package:crowdleague/enums/auth_step.dart';
+import 'package:crowdleague/enums/auto_validate.dart';
+import 'package:crowdleague/enums/email_auth_mode.dart';
+import 'package:crowdleague/models/app/app_state.dart';
+import 'package:crowdleague/reducers/auth/update_email_auth_options_page.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('UpdateEmailAuthOptionsPageReducer', () {
+    test('updates store.emailAuthOptionsPage.mode', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:11]: cool beans

File: test/unit/reducers/update_email_auth_options_page_test.dart [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+      // Setup an initial app state
+      final initialState = AppState.init();
+
+      // Create the reducer.
+      final rut = UpdateEmailAuthOptionsPageReducer();
+
+      // Create data to send with action
+      final testSignUpMode = EmailAuthMode.signUp;
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:19]: can do,thought it was clearer putting it in a variable, but the enum is super clear on its own, dont need no var

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:19]: updated :)

File: test/unit/reducers/update_email_auth_options_page_test.dart [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+      final testSignInMode = EmailAuthMode.signIn;
+
+      // Invoke the reducer to get a new state.
+      final newStateSignUp = rut.reducer(
+          initialState, UpdateEmailAuthOptionsPage(mode: testSignUpMode));
+
+      // Check that the new state has the new user.
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:26]: updated

File: test/unit/reducers/update_email_auth_options_page_test.dart [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+      expect(newStateSignUp.emailAuthOptionsPage.mode, testSignUpMode);
+
+      // Invoke the reducer to get a new state.
+      final newStateSignIn = rut.reducer(
+          initialState, UpdateEmailAuthOptionsPage(mode: testSignInMode));
+
+      // Check that the new state has the new user.
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:33]: updated

File: test/unit/reducers/update_email_auth_options_page_test.dart [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+import 'package:test/test.dart';
+
+void main() {
+  group('UpdateEmailAuthOptionsPageReducer', () {
+    test('updates store.emailAuthOptionsPage.mode', () {
+      // Setup an initial app state
+      final initialState = AppState.init();
+
+      // Create the reducer.
+      final rut = UpdateEmailAuthOptionsPageReducer();
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:16]: yup :)

File: test/unit/reducers/update_email_auth_options_page_test.dart [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+      // Check that the new state has the new user.
+      expect(newStateSignUp.emailAuthOptionsPage.mode, testSignUpMode);
+
+      // Invoke the reducer to get a new state.
+      final newStateSignIn = rut.reducer(
+          initialState, UpdateEmailAuthOptionsPage(mode: testSignInMode));
+
+      // Check that the new state has the new user.
+      expect(newStateSignIn.emailAuthOptionsPage.mode, testSignInMode);
+    });
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:35]: addoid

File: test/unit/reducers/update_email_auth_options_page_test.dart [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+    test('updates store.emailAuthOptionsPage.showPassword', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/reducers/update_email_auth_options_page_test.dart:36]: yup

File: test/unit/reducers/store_user_test.dart (hunk 18 above)
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/reducers/store_user_test.dart:15]: awesome!  glad you like it 👍

File: test/unit/reducers/store_user_test.dart (hunk 17 above)
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/reducers/store_user_test.dart:24]: 🤣

**Review by nickmeinhold — APPROVED** [review]
Thanks the great PR!  And thanks for writing all these awesome tests... it's so great seeing the numbers of tests that get run when we run a `flutter test` to see if new changes have broken anything 🙂

File: test/unit/reducers/store_auth_step_test.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/reducers/store_auth_step_test.dart]: Hahaha I actually didn't see this when I was going through earlier... so glad I skimmed the email later 😄

File: test/unit/reducers/store_auth_step_test.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/reducers/store_auth_step_test.dart]: 🙂

File: test/unit/reducers/update_email_auth_options_page_test.dart (hunk 23 above)
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply test/unit/reducers/update_email_auth_options_page_test.dart]: Nice!  Yeah there's definitely a place for the variable adding clarity hey

**Pull request merged**